### ISSUE TITLE
Failed to Reproduce AquaLoRA on SD2.1 base

### ISSUE THREAD

**Issue by XiaoxiKong** [Georgefwt/AquaLoRA#16]
Hi authors,

Thank you so much for sharing your code and making it available to the research community. We have been working on reproducing your results on Stable Diffusion 2.1 base, the first stage is grate, but we've encountered a few technical challenges that we'd like to ask for your guidance on.

### Environment
- `diffusers` version: 0.31.0
- `peft` version: 0.14.0

### Modifications Made

**1. Add replace_unet_linears_with_lora function:**

To enable LoRA functionality in our environment (converting standard nn.Linear layers to LoRACompatibleLinear), we implemented a function to process the unet_keys and wrap the corresponding layers so they can be fine-tuned. Our implementation is as follows:
```python
def replace_unet_linears_with_lora(unet, unet_keys):
    for key in unet_keys:
        module = unet
        sub_keys = key.split(".")
        for sub_key in sub_keys[:-1]:
            module = getattr(module, sub_key)
        last_key = sub_keys[-1]
        orig_layer = getattr(module, last_key)
        
        if isinstance(orig_layer, torch.nn.Linear):
            new_module = LoRACompatibleLinear(orig_layer.in_features, orig_layer.out_features)
            new_module.weight.data = orig_layer.weight.data.clone()
            if orig_layer.bias is not None:
                new_module.bias.data = orig_layer.bias.data.clone()
            setattr(module, last_key, new_module)
```
This function processes the specified unet_keys to convert the original linear layers into LoRACompatibleLinear layers that can be properly fine-tuned.

**2. Workaround for `mapped_loradiag` parameter:**

During fine-tuning in second stage, we encountered an issue where the diagonal matrix could not be passed correctly to the LoRA layer.
In the original code, the UNet forward call was written as:mapped_loradiag.
``` Python:
model_pred = unet(
    noisy_wm_model_input,
    timesteps,
    encoder_hidden_states,
    class_labels=class_labels,
    cross_attention_kwargs={"scale": mapped_loradiag}
).sample ```

However, the parameter was not successfully propagated into the function inside the LoRA layers. As a temporary workaround, we stored as a global variable and manually passed it into the function during execution.

### Question

Could you help clarify why the `mapped_loradiag` parameter cannot be passed normally? Are there any recommended solutions or did we miss something in the setup?
Thank you again for your contributions!

**Comment by Georgefwt** [comment]
First, I have to clarify that I replaced the LoRA for the Linear layer. Please check here:  
https://github.com/Georgefwt/AquaLoRA/blob/62fb5cf186050dd4e3bce0bfe15068872fc57d62/train/ppft_train.py#L659  
https://github.com/Georgefwt/AquaLoRA/blob/62fb5cf186050dd4e3bce0bfe15068872fc57d62/train/ppft_train.py#L680  

For question 2, since I'm not entirely sure what changes you made, I can't provide specific suggestions. However, the current version should be working fine. Please double-check your code implementation.  

Also, note that SD2 uses v-prediction. I have already implemented the PPFT loss for the v-prediction version in the code. Please refer to:  
https://github.com/Georgefwt/AquaLoRA/blob/62fb5cf186050dd4e3bce0bfe15068872fc57d62/train/ppft_train.py#L1047

**Comment by XiaoxiKong** [comment]
**1.
Thank you very much for the clarification!
We noticed that in this part of the code, a LoRA layer is attached to the attn_processor module:
```python
elif isinstance(attn_processor, LoRACompatibleLinear):
    ctx = nullcontext
    with ctx():
        lora = LoRALinearLayer(
            attn_processor.in_features,
            attn_processor.out_features,
            rank,
        )
```
However, during runtime, an error occurs indicating that attn_processor is not an instance of LoRACompatibleLinear.
We suspect that this issue might be related to a version mismatch between the peft and diffusers packages.
To make the code run properly, we manually modified the modules corresponding to unet_keys to be instances of LoRACompatibleLinear.
Could you please confirm if this behavior is expected or if we might be missing something?
2.
In our implementation, we slightly modified the design to dynamically pass the LoRA scaling during training. Specifically, we define a global dictionary and update it at every training step as follows:GLOBAL_LORA_SCALE = {}
```python
for name, module in unet.named_modules():
    if hasattr(module, "lora_layer") and module.lora_layer is not None:
        GLOBAL_LORA_SCALE[id(module.lora_layer)] = mapped_loradiag
```
Then, in our modified version of CustomLoRALinearLayer.forward, we fetch the corresponding scale using:
```python
scale = kwargs.get("scale", GLOBAL_LORA_SCALE.get(id(self), 1.0))
```
This approach allows us to dynamically assign per-layer LoRA scaling values at each step.
Could you please confirm whether this design is consistent with your intended implementation, if we might be missing something?
We’ll also carefully review our own code as you suggested. We did notice that we hadn’t correctly set the PPFT loss for the v-prediction version in SD2, as we hadn’t carefully checked that part before.
Thank you again for your helpful reminder and guidance!

**Comment by Georgefwt** [comment]
You are welcome.

**For Question 1:**
My suggestion is to use the package versions specified in the `requirements.txt`. It’s quite difficult to debug remotely if the code is modified by you own. I just tested it myself, and the environment can be installed correctly according to the README. Also, `diffusers` version 0.24 does support SD2.1.

**For Question 2:**
I don’t think we need *per-layer* LoRA scaling values -- the scaling values should be the same across all layers. However, different samples within a batch can have different scaling values.
At first glance, your implementation looks fine, but I'm a bit concerned about whether it handles batches properly. Wouldn't your current implementation result in all samples in a batch sharing the same scaling value?

**Comment by XiaoxiKong** [comment]
Hi, thanks for your help!
The issue was indeed caused by environment inconsistencies on my side. After switching to the same environment as described in the repo, the model trained successfully.
I really appreciate your guidance and for sharing this great work!

**Comment by Georgefwt** [comment]
You are welcome!
